server/src/main/java/org/elasticsearch/transport/OutboundHandler.java

```diff
@@ -241,10 +260,101 @@ private void sendMessage(
         );
     }
 
+    // public for tests
+    public static BytesReference serialize(
+        @Nullable String requestAction,
```

Comment:
I get your point @DaveCTurner, setting request or response based on a null is not great, but then again, this method already has an absurd number of parameters? :)

Reply:
Maybe it needs a parameters object? I think I'd call it... `OutboundMessage` 🤣

Reply:
I'll follow up with something like this:

https://github.com/original-brownbear/elasticsearch/compare/stop-retaining-transport-response-past-sending...DaveCTurner:elasticsearch:2025/03/21/OutboundHandler-MessageDirection

drops the `isError` boolean that only makes sense on responses, replacing it with a three-state enum, and makes the `action` available on all paths.